File: leetcode/p543.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    pair<int, int> dfs(TreeNode* root) {
        pair<int, int> left = {0, 0};
        pair<int, int> right = {0, 0};
        if (root->left) {
            left = dfs(root->left);
        }
        if (root->right) {
            right = dfs(root->right);
        }
        int max_d = left.first > right.first ? left.first : right.first;
        max_d = left.second+right.second > max_d ? left.second+right.second : max_d;
        int max_dep = left.second > right.second ? left.second+1 : right.second+1;
        return {max_d, max_dep};
    }
    int diameterOfBinaryTree(TreeNode* root) {
        if (root == NULL) return 0;
        pair<int, int> res = dfs(root);
        return res.first;
    }
};

File: leetcode/p543.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, node):
        deep_l = 0
        max_l = 0
        deep_r = 0
        max_r = 0
        if node.left:
            deep_l, max_l = self.dfs(node.left)
            deep_l += 1
        if node.right:
            deep_r, max_r = self.dfs(node.right)
            deep_r += 1
        #print(node.val, deep_l, deep_r, max_l, max_r)
        return max(deep_l, deep_r), max(deep_l+deep_r, max_l, max_r)

    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        if not root:
            return 0
        _, max_path = self.dfs(root)
        return max_path
